### PR TITLE
KAFKA-13907: Fix hanging ServerShutdownTest.testCleanShutdownWithKRaftControllerUnavailable

### PR DESCRIPTION
**Pull request by soarez**
When a controlled shutdown is requested the broker tries to communicate the state change to the controller via a heartbeat request. [1]

In this test, the controller is not available so the request will fail. The current timeout behavior in a heartbeat request is to just keep retrying — which generally makes sense, just not in the context of a controlled shutdown.

When a heartbeat request times out, if we are in the middle of a controlled shutdown, we shouldn't just retry forever but rather just give up on trying to contact the controller and proceed with the controlled shutdown.

[1] https://github.com/apache/kafka/blob/f2d6282668a31b9a554563338f9178e2bba2833f/core/src/main/scala/kafka/server/BrokerLifecycleManager.scala#L217

*Summary of testing strategy*
The test no longer fails

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cmccabe** [comment]
Sorry, but this is not correct. We can't proceed with the controlled shutdown if the controller is not available.

It looks like the timeout needs to be adjusted in the test since the behavior we want is to time out and do a dirty shutdown.

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala [hunk 1]
```diff
@@ -431,8 +431,18 @@ class BrokerLifecycleManager(val config: KafkaConfig,
     }
 
     override def onTimeout(): Unit = {
-      info("Unable to send a heartbeat because the RPC got timed out before it could be sent.")
-      scheduleNextCommunicationAfterFailure()
+      if (_state == BrokerState.PENDING_CONTROLLED_SHUTDOWN) {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/BrokerLifecycleManager.scala]: I think one way we can fix this test is to let `KafkaBroker.shutdown` accept a timeout parameter. Possibly we could overload it like this:
```
def shutdown(timeoutMs: Long): Unit
def shutdown(): Unit = shutdown(5000)
```
Then we can change the test to just wait a couple seconds. What do you think?

Comment (soarez) [reply core/src/main/scala/kafka/server/BrokerLifecycleManager.scala]: If we modify the interface in this way it looks like the timeout refers to the overall shutdown operation, but the timeout is specific to the controlled shutdown attempt.

Since the test applies only to KRaft we can cast the broker to the known type (`BrokerServer`) and modify a new _VisibleForTesting_ variable that holds the default timeout. WDYT?

**Comment by soarez** [comment]
Thanks for the review @cmccabe and @hachikuji. 

I now realize my initial proposal is silly. A controlled shutdown requires gracefully moving leadership away from the broker, so we simply cannot do it if the controller is unreachable.

I have updated the PR to adjust the timeout for the test instead. Please take another look and share any thoughts.

File: core/src/main/scala/kafka/server/BrokerServer.scala [hunk 2]
```diff
@@ -93,6 +93,9 @@ class BrokerServer(
 
   this.logIdent = logContext.logPrefix
 
+  // VisibleForTesting
+  private[kafka] var controlledShutdownTimeoutMillis: Long = 5 * 60 * 1000L // 5 minutes
```
Comment (dengziming) [inline core/src/main/scala/kafka/server/BrokerServer.scala]: This can solve the problem but is not very elegant, we can add a `shutdown(timeout)`, and use `Time` to compute a deadline, then we can use the deadline to compute a timeout for a controlled shutdown, WDYT?

Comment (soarez) [reply core/src/main/scala/kafka/server/BrokerServer.scala]: Thanks for reviewing this @dengziming. I agree it is not an elegant approach.
And thanks for your suggestion, it seems to be in line with @hachikuji 's suggestion as well.
I've updated this patch to follow your suggested approach. Please let me know if this is what you had in mind.
However, I'm not very confident that this is better, since `KafkaServer` also has to implement the new `shutdown` variant and it has no use for the timeout. So it seems a bit strange to have an implementation where the timeout has to be ignored.

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala [hunk 3]
```diff
@@ -248,9 +248,11 @@ abstract class KafkaServerTestHarness extends QuorumTestHarness {
     index
   }
 
-  def killBroker(index: Int): Unit = {
+  def killBroker(index: Int): Unit = killBroker(index, TimeUnit.SECONDS.toMinutes(5))
```
Comment (dengziming) [inline core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala]: Thanks for the update, can we unify these 2 methods into one: 
`def killBroker(index: Int, timeoutMs: Long = TimeUnit.SECONDS.toMinutes(5)): Unit = {...}`

Comment (soarez) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala]: Thanks for giving it another look. I didn't initially do this because the method is also called from Java, where default parameters are not supported. However, it's just one occurrence, so I've updated the caller in that case. I also noticed that `toMinutes` here should be `toMillis` instead.

**Review by dengziming — APPROVED** [review]
LGTM

File: core/src/main/scala/kafka/server/KafkaBroker.scala [hunk 4]
```diff
@@ -93,6 +93,7 @@ trait KafkaBroker extends Logging {
   def startup(): Unit
   def awaitShutdown(): Unit
   def shutdown(): Unit
+  def shutdown(timeoutMs: Long): Unit
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaBroker.scala]: How about using `Duration` instead of long type? Also, we can rename it from `timeoutMs` to `timeout`

File: core/src/main/scala/kafka/server/BrokerServer.scala [hunk 5]
```diff
@@ -623,9 +623,12 @@ class BrokerServer(
     }
   }
 
-  override def shutdown(): Unit = {
+  override def shutdown(): Unit = shutdown(TimeUnit.MINUTES.toMillis(5))
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/BrokerServer.scala]: How about adding default implementation to parent class?

**Review by chia7712 — COMMENTED** [review]
@soarez thanks for updating PR. two minor comments left

File: core/src/main/scala/kafka/server/KafkaBroker.scala (hunk 4 above)
Comment (soarez) [reply core/src/main/scala/kafka/server/KafkaBroker.scala]: Good suggestion. Applied

File: core/src/main/scala/kafka/server/BrokerServer.scala (hunk 5 above)
Comment (soarez) [reply core/src/main/scala/kafka/server/BrokerServer.scala]: Yup, that makes sense. Changed

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala [hunk 6]
```diff
@@ -260,9 +274,12 @@ abstract class KafkaServerTestHarness extends QuorumTestHarness {
   }
 
   def killBroker(index: Int): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala:267]: maybe we should keep origin implementation since it expect to await shutdown.

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala [hunk 7]
```diff
@@ -260,9 +274,12 @@ abstract class KafkaServerTestHarness extends QuorumTestHarness {
...
-    if (alive(index)) {
-      _brokers(index).shutdown()
-      _brokers(index).awaitShutdown()
+    killBroker(index, Duration.ofSeconds(5))
+  }
+
+  def killBroker(index: Int, timeout: Duration): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala:275]: we need to document the difference of this variety.

File: core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala [hunk 8]
```diff
@@ -197,12 +198,11 @@ class ServerShutdownTest extends KafkaServerTestHarness {
     verifyNonDaemonThreadsStatus()
   }
 
-  @Disabled
   @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
   @ValueSource(strings = Array("kraft"))
   def testCleanShutdownWithKRaftControllerUnavailable(quorum: String): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala]: the shutdown with timeout is a kind of dirty shutdown so we should rename the test to `testDirtyShutdownWithKRaftControllerUnavailable`

File: core/src/test/java/kafka/test/junit/ZkClusterInvocationContext.java [hunk 9]
```diff
@@ -323,7 +324,7 @@ public void rollingBrokerRestart() {
                 throw new IllegalStateException("Tried to restart brokers but the cluster has not been started!");
             }
             for (int i = 0; i < clusterReference.get().brokerCount(); i++) {
-                clusterReference.get().killBroker(i);
+                clusterReference.get().killBroker(i, Duration.ofSeconds(5));
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ZkClusterInvocationContext.java]: what is the purpose of this change?

**Review by chia7712 — COMMENTED** [review]
@soarez thanks for updating code. please take a look at those comments. thanks!

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala (hunk 6 above)
Comment (soarez) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala:267]: The original implementation has a timeout of 5 minutes, so I'll keep the same behavior here.

File: core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala (hunk 8 above)
Comment (soarez) [reply core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala]: That's a good point. If the controller is unavailable, the shutdown can't really be considered clean. However, I think it's better to just remove the `Clean` word from the method, rather than replacing it with `Dirty`. It seems cleaner and less confusing, as it might not be clear what "dirty" is referring to. Please let me know if you disagree.

File: core/src/test/java/kafka/test/junit/ZkClusterInvocationContext.java (hunk 9 above)
Comment (soarez) [reply core/src/test/java/kafka/test/junit/ZkClusterInvocationContext.java]: I think this is unnecessary. Removing

File: core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala (hunk 8 above)
Comment (chia7712) [reply core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala]: That is good to me

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala (hunk 6 above)
Comment (chia7712) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala:267]: My point was "should we keep `_brokers(index).awaitShutdown()`"? It seems the latch used by `awaitShutdown` should be notified after `shutdown` get completed, so maybe it is fine to remove it. It would be great to add comments if you prefer to remove `awaitShutdown`

Comment (soarez) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala:267]: Oh I see. It was a mistake to remove `awaitShutdown`. I think we should leave the behavior as is, I don't see any reason why the new `killBroker(idx, timeout)` alternative shouldn't also call `awaitShutdown`.

**Comment by chia7712** [comment]
@soarez Could you please fix the conflicts? thanks!

**Comment by soarez** [comment]
@chia7712 done. It seems to me that the current test failures are unrelated. Could you confirm?

**Comment by chia7712** [comment]
@soarez I believe those failed tests are unrelated, but could you rebase code to trigger QA again due to incompletion of `JDK 17 and Scala 2.13`?
